Documents/FileSystemSpec.md

```diff
@@ -1,2 +1,600 @@
+
+_Server Capabilities:_ See general file system provider [server capabilities](#fileSystemProviderServerCapabilities)
+
+_Notification:_
+- method: `fileSystem/didChangeFile`
+- params: `DidChangeFileParams` defined as follows:
+
+```typescript
+/**
+ * An event to signal that a resource has been created, changed, or deleted. This
+ * event should fire for resources that are being [watched](#FileSystemProvider.watch)
+ * by clients of this provider.
+ *
+ * *Note:* It is important that the metadata of the file that changed provides an
+ * updated `mtime` that advanced from the previous value in the [stat](#FileStat) and a
+ * correct `size` value. Otherwise there may be optimizations in place that will not show
+ * the change in an editor for example.
+ */
+export interface DidChangeFileParams {
+    /**
+     * The change events.'
+     */
+    changes: FileChangeEvent[];
+}
+
+/**
+ * The event filesystem providers must use to signal a file change.
+ */
+export interface FileChangeEvent {
+    /**
+     * The type of change.
+     */
+    uri: URI;
+
+    /**
+     * The uri of the file that has changed.
+     */
+    type: FileChangeType
+}
+
+/**
+ * Enumeration of file change types.
+ */
+export namespace FileChangeType {
+    /**
+     * The contents or metadata of a file have changed.
+     */
+    export const Changed = 1;
+
+    /**
+     * A file has been created.
+     */
+    export const Created = 2;
+
+    /**
+     * A file has been deleted.
+     */
+    export const Deleted = 3;
+}
+```
+
+#### <a href="#watch" name="watch" class="anchor">Watch Request (:arrow_right_hook:)</a>
+
+Watch requests are sent from the client to the server to subscribe to [DidChangeFile](#didChangeFile) events in the file or folder denoted by `uri`.
+
+_Client Capabilities:_ See general file system provider [client capabilities](#fileSystemProviderClientCapabilities)
+
+_Server Capabilities:_ See general file system provider [server capabilities](#fileSystemProviderServerCapabilities)
+
+_Request:_
+- method: `fileSystem/watch`
+- params: `WatchParams` defined as follows:
+
+
+```typescript
+export interface WatchParams {
+    /**
+     * The uri of the file or folder to be watched.
+     */
+    uri: URI;
+
+    /**
+     * Configures the watch
+     */
+    options: WatchFileOptions
+}
+
+export interface WatchFileOptions {
+    /**
+     * If a folder should be recursively subscribed to
+     */
+    recursive: boolean;
+
+    /**
+     * Folders or files to exclude from being watched.
+     */
+    excludes: string[];
+}
+```
+
+_Response:_
+- result: `WatchResponse`
+
+```typescript
+export interface WatchResponse {
+    /**
+     * The subscription ID to be used in order to stop watching a file or folder via the [StopWatching](#stopWatching) notification.
+     */
+    subscriptionId: number;
+}
+```
+
+#### <a href="#stopWatching" name="stopWatching" class="anchor">StopWatching Notification (:arrow_right:)</a>
```

Comment:
@dbaeumer This was an interesting fallout from adding `watch`. Needed a way to `unwatch` so built the idea of a `subscriptionId` that can be passed back to stop watching.

Reply:
In LSP we follow the concept that the sender of the message generates an id. This has the advantage that if something goes wrong the sender can still un-register since it knows the id. See also https://microsoft.github.io/language-server-protocol/specifications/specification-current/#client_registerCapability

Reply:
Good idea! Updated.